File: 40-combination-sum-ii/combination-sum-ii.py
"""

10,1,2,7,6,1,5

1,1,2,5,6,7,10

"""
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        def REC(index, total, arr):
            if total == target:
                result.append(arr.copy())
                return
            if index == len(candidates) or total > target:
                return
            #Pick
            arr.append(candidates[index])
            REC(index + 1, total + candidates[index], arr)
            arr.pop()
            #Do not Pick
            while index + 1 < len(candidates) and candidates[index + 1] == candidates[index]:
                index+=1
            REC(index + 1, total, arr)
            
        candidates.sort()  
        result = []
        REC(0,0,[])
        return result

File: 40-combination-sum-ii/combination-sum-ii.java
class Solution {
    public List<List<Integer>> result;
    public int targetG;
    public int [] candidatesG;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        result = new ArrayList<>();
        targetG = target;
        Arrays.sort(candidates);
        candidatesG = candidates;
        fun(0,new ArrayList<>(),0);
        return result;
    }
    public void fun(int index, List<Integer> curr, int sum){
        if(sum == targetG){
            result.add(new ArrayList<>(curr));
            return;
        }
        if(sum > targetG || index == candidatesG.length){
            return;
        }
        curr.add(candidatesG[index]);
        fun(index + 1, curr, sum + candidatesG[index]);
        while(index + 1 < candidatesG.length && candidatesG[index] == candidatesG[index + 1]){
            index++;
        }
        curr.remove(curr.size()-1);
        fun(index + 1,curr,sum);
    }
}